Fix warnings in clang 10 build

After the last expression template update, some warnings popped up in the clang 10 build.

```
/usr/workspace/bramwell/gitspaces/serac/src/physics/../numerics/vector_expression.hpp:119:75: warning: operator '?:' has lower precedence than '+'; '+' will be evaluated first [-Wparentheses]
  const long long per_proc = (SIZE / num_procs) + (SIZE % num_procs != 0) ? 1 : 0;
                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
/usr/workspace/bramwell/gitspaces/serac/tests/expr_templates.cpp:62:3: note: in instantiation of function template specialization 'serac::evaluate<serac::internal::BinaryVectorExpr<mfem::Vector &, mfem::Vector &, std::plus<double> > >' requested here
  evaluate(lhs + rhs, expr_result, MPI_COMM_WORLD);
  ^
/usr/workspace/bramwell/gitspaces/serac/src/physics/../numerics/vector_expression.hpp:119:75: note: place parentheses around the '+' expression to silence this warning
  const long long per_proc = (SIZE / num_procs) + (SIZE % num_procs != 0) ? 1 : 0;
                                                                          ^
                             (                                           )
/usr/workspace/bramwell/gitspaces/serac/src/physics/../numerics/vector_expression.hpp:119:75: note: place parentheses around the '?:' expression to evaluate it first
  const long long per_proc = (SIZE / num_procs) + (SIZE % num_procs != 0) ? 1 : 0;
                                                                          ^
                                                  (                              )
```

--- a/src/docs/sphinx/expr_templates.rst
+++ b/src/docs/sphinx/expr_templates.rst
@@ -26,9 +26,11 @@ In particular, Serac currently provides the following operations:
 5. Application of an ``mfem::Operator`` with ``op * a`` for ``mfem::Operator op`` and vector ``a`` - note that 
     because ``mfem::Matrix`` inherits from ``mfem::Operator`` this functionality includes matrix-vector multiplication.
 
-
 Note that all of these expressions can be composed, that is, an expression can be used as an argument
 to another expression.  This allows for chaining, e.g., ``-a + b + 0.3 * c``.
+
+Memory Safety
+-------------
 
 Because these expression objects are imperfect closures (as with lambdas in C++), care should be taken to
 ensure that objects are not used after they go out of scope.  
@@ -76,3 +78,29 @@ As with the previous example, this should be resolved by moving the ``mfem::Vect
         auto a3 = evaluate(a * 3.0); // a is of type mfem::Vector
         return std::move(a3) - b; // return value is of expression type
     };
+
+
+Avoiding Allocations
+--------------------
+
+Currently, assigning an expression template to a vector, e.g. ``mfem::Vector c = a + b``
+will result in an allocation.  To avoid this, use the ``evaluate`` function:
+
+.. code-block:: cpp
+
+    // Preallocate
+    mfem::Vector c(size);
+    evaluate(a + b, c);
+
+
+
+Distributed Expression Templates
+--------------------------------
+
+.. note::
+    This functionality is under active development and may change significantly.
+
+Distributed expression templates (with ``mfem::HypreParVector``) are **experimentally** supported. 
+Mixed operations (where some operands are global ``mfem::Vector``s and others are distributed vectors)
+are not supported.  Additionally, an expression cannot be assigned to a ``HypreParVector``.  Use the 
+``evaluate`` function described above.

--- a/src/numerics/expr_template_internal.hpp
+++ b/src/numerics/expr_template_internal.hpp
@@ -27,11 +27,14 @@ namespace serac::internal {
  * @brief Determines whether a given type should be owned by a vector expression
  * @tparam vec The vector expression type
  * @note A vector should be owned if it is not an (optionally const) reference
- * to an mfem::Vector.  Note that std::remove_const does not apply here because
- * const is not a top-level qualifier on a const reference
+ * to an mfem::Vector or a class derived from mfem::Vector.  Note that
+ * std::remove_const does not apply here because const is not a top-level
+ * qualifier on a const reference
  */
 template <typename vec>
-inline constexpr bool owns_v = !std::is_same_v<vec, const mfem::Vector&> && !std::is_same_v<vec, mfem::Vector&>;
+inline constexpr bool owns_v =
+    !(std::is_base_of_v<mfem::Vector, std::decay_t<vec>> &&
+      (std::is_same_v<vec, const std::decay_t<vec>&> || std::is_same_v<vec, std::decay_t<vec>&>));
 
 /**
  * @brief Type alias for what should be stored by a vector expression - an object
@@ -49,6 +52,26 @@ using vec_t = typename std::conditional_t<owns_v<vec>, std::decay_t<vec>, const 
  */
 template <typename vec>
 using vec_arg_t = typename std::conditional_t<owns_v<vec>, std::decay_t<vec>&&, const vec&>;
+
+/**
+ * @brief Wraps the indexing of a vector type
+ * @param[in] v The vector to index
+ * @param[in] idx The offset of the indexing
+ * @note This is needed for uniform indexing that includes mfem::HypreParVectors.
+ * Because mfem::Vector::operator[] is implemented as an implicit conversion to double*,
+ * the additional implicit conversion (mfem::HypreParVector::operator hypre_ParVector*)
+ * results in an ambiguous call to operator[], as both implicit conversions are pointer
+ * types on which pointer arithmetic can be performed.
+ */
+template <typename vec>
+auto index(vec&& v, const std::size_t idx)
+{
+  if constexpr (std::is_same_v<std::decay_t<vec>, mfem::HypreParVector>) {
+    return static_cast<const double*>(v)[idx];
+  } else {
+    return v[idx];
+  }
+}
 
 /**
  * @brief Derived VectorExpr class for representing the application of a unary
@@ -72,7 +95,7 @@ public:
    * expression at index @p i
    * @param i The index to evaluate at
    */
-  double operator[](size_t i) const { return op_(v_[i]); }
+  double operator[](size_t i) const { return op_(index(v_, i)); }
   /**
    * @brief Returns the size of the vector expression
    */
@@ -168,7 +191,7 @@ public:
    * expression at index @p i
    * @param i The index to evaluate at
    */
-  double operator[](size_t i) const { return op_(u_[i], v_[i]); }
+  double operator[](size_t i) const { return op_(index(u_, i), index(v_, i)); }
   /**
    * @brief Returns the size of the vector expression
    */

--- a/src/numerics/expr_template_ops.hpp
+++ b/src/numerics/expr_template_ops.hpp
@@ -17,10 +17,10 @@
 
 /**
  * @brief A utility for conditionally enabling templates if a given type
- * is an mfem::Vector
+ * is an mfem::Vector (including by inheritance)
  */
 template <typename MFEMVec>
-using enable_if_mfem_vec = std::enable_if_t<std::is_same_v<std::decay_t<MFEMVec>, mfem::Vector>>;
+using enable_if_mfem_vec = std::enable_if_t<std::is_base_of_v<mfem::Vector, std::decay_t<MFEMVec>>>;
 
 template <typename T>
 auto operator-(serac::VectorExpr<T>&& u)

--- a/src/numerics/vector_expression.hpp
+++ b/src/numerics/vector_expression.hpp
@@ -88,8 +88,10 @@ void evaluate(const VectorExpr<T>& expr, mfem::Vector& result)
 {
   SLIC_ERROR_IF(expr.Size() != static_cast<std::size_t>(result.Size()),
                 "Vector sizes in expression assignment must be equal");
+  // Get the underlying array for indexing compatibility with mfem::HypreParVector
+  double* result_arr = result;
   for (size_t i = 0; i < expr.Size(); i++) {
-    result[i] = expr[i];
+    result_arr[i] = expr[i];
   }
 }
 
@@ -116,7 +118,7 @@ void evaluate(const VectorExpr<T>& expr, mfem::Vector& result, MPI_Comm comm)
   double* result_arr = result;
 
   // If the array size is not divisible by # elements, add one
-  const long long per_proc = (SIZE / num_procs) + (SIZE % num_procs != 0) ? 1 : 0;
+  const long long per_proc = (SIZE / num_procs) + ((SIZE % num_procs != 0) ? 1 : 0);
 
   // Truncate the number of elements for the last process
   const long long n_entries = (rank == num_procs - 1) ? SIZE - ((num_procs - 1) * per_proc) : per_proc;

--- a/tests/benchmark_expr_templates.cpp
+++ b/tests/benchmark_expr_templates.cpp
@@ -34,6 +34,29 @@ static std::pair<mfem::DenseMatrix, mfem::Vector> sample_matvec(const int rows, 
   return {matrix, vec_in};
 }
 
+static auto build_partitioning(MPI_Comm comm, const int size)
+{
+  int num_procs = 0;
+  int rank      = 0;
+  MPI_Comm_size(comm, &num_procs);
+  MPI_Comm_rank(comm, &rank);
+  bool assumed_partition = HYPRE_AssumedPartitionCheck();
+  auto partitioning      = std::make_unique<int[]>(assumed_partition ? 2 : (num_procs + 1));
+  auto per_proc          = (size / num_procs) + ((size % num_procs != 0) ? 1 : 0);
+
+  if (assumed_partition) {
+    auto n_entries  = (rank == num_procs - 1) ? size - ((num_procs - 1) * per_proc) : per_proc;
+    partitioning[0] = per_proc * rank;
+    partitioning[1] = (per_proc * rank) + n_entries;
+  } else {
+    for (int i = 0; i < num_procs; i++) {
+      partitioning[i] = per_proc * i;
+    }
+    partitioning[num_procs] = size;
+  }
+  return std::make_pair(std::move(partitioning), per_proc * rank);
+}
+
 static void BM_mixed_expr_MFEM(benchmark::State& state)
 {
   MPI_Barrier(MPI_COMM_WORLD);
@@ -178,12 +201,38 @@ static void BM_large_expr_single_alloc_par_EXPR(benchmark::State& state)
   MPI_Barrier(MPI_COMM_WORLD);
 }
 
+static void BM_large_expr_single_alloc_hypre_par_EXPR(benchmark::State& state)
+{
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  // Number of rows is the argument that varies
+  const int rows  = state.range(0);
+  auto [lhs, rhs] = sample_vectors(rows);
+
+  auto [partitioning, start] = build_partitioning(MPI_COMM_WORLD, rows);
+
+  mfem::HypreParVector lhs_par(MPI_COMM_WORLD, rows, lhs + start, partitioning.get());
+  mfem::HypreParVector rhs_par(MPI_COMM_WORLD, rows, rhs + start, partitioning.get());
+
+  mfem::HypreParVector expr_result(MPI_COMM_WORLD, rows, partitioning.get());
+
+  for (auto _ : state) {
+    // This code gets timed
+    evaluate(lhs + rhs + lhs + rhs + lhs + rhs + lhs + rhs + lhs + rhs + lhs + rhs + lhs + rhs, expr_result);
+  }
+
+  MPI_Barrier(MPI_COMM_WORLD);
+}
+
 BENCHMARK(BM_mixed_expr_MFEM)->RangeMultiplier(2)->Range(10, 10 << 10);
 BENCHMARK(BM_mixed_expr_EXPR)->RangeMultiplier(2)->Range(10, 10 << 10);
 BENCHMARK(BM_mixed_expr_single_alloc_EXPR)->RangeMultiplier(2)->Range(10, 10 << 10);
 BENCHMARK(BM_large_expr_MFEM)->RangeMultiplier(2)->Range(10, 10 << 10);
 BENCHMARK(BM_large_expr_single_alloc_EXPR)->RangeMultiplier(2)->Range(10, 10 << 10);
-BENCHMARK(BM_large_expr_single_alloc_par_EXPR)->RangeMultiplier(2)->Range(10, 10 << 10);
+
+// Too slow
+BENCHMARK(BM_large_expr_single_alloc_par_EXPR)->RangeMultiplier(2)->Range(10, 10 << 5);
+BENCHMARK(BM_large_expr_single_alloc_hypre_par_EXPR)->RangeMultiplier(2)->Range(10, 10 << 10);
 
 //------------------------------------------------------------------------------
 #include "axom/slic/core/UnitTestLogger.hpp"

--- a/tests/expr_templates.cpp
+++ b/tests/expr_templates.cpp
@@ -32,6 +32,29 @@ static std::pair<mfem::DenseMatrix, mfem::Vector> sample_matvec(const int rows, 
   return {matrix, vec_in};
 }
 
+static auto build_partitioning(MPI_Comm comm, const int size)
+{
+  int num_procs = 0;
+  int rank      = 0;
+  MPI_Comm_size(comm, &num_procs);
+  MPI_Comm_rank(comm, &rank);
+  bool assumed_partition = HYPRE_AssumedPartitionCheck();
+  auto partitioning      = std::make_unique<int[]>(assumed_partition ? 2 : (num_procs + 1));
+  auto per_proc          = (size / num_procs) + ((size % num_procs != 0) ? 1 : 0);
+
+  if (assumed_partition) {
+    auto n_entries  = (rank == num_procs - 1) ? size - ((num_procs - 1) * per_proc) : per_proc;
+    partitioning[0] = per_proc * rank;
+    partitioning[1] = (per_proc * rank) + n_entries;
+  } else {
+    for (int i = 0; i < num_procs; i++) {
+      partitioning[i] = per_proc * i;
+    }
+    partitioning[num_procs] = size;
+  }
+  return std::make_pair(std::move(partitioning), per_proc * rank);
+}
+
 TEST(expr_templates, basic_add)
 {
   MPI_Barrier(MPI_COMM_WORLD);
@@ -64,6 +87,33 @@ TEST(expr_templates, basic_add_parallel)
   for (int i = 0; i < size; i++) {
     EXPECT_FLOAT_EQ(mfem_result[i], expr_result[i]);
   }
+  MPI_Barrier(MPI_COMM_WORLD);
+}
+
+TEST(expr_templates, basic_add_hyprepar)
+{
+  MPI_Barrier(MPI_COMM_WORLD);
+  constexpr int size = 10;
+  auto [lhs, rhs]    = sample_vectors(size);
+
+  auto [partitioning, start] = build_partitioning(MPI_COMM_WORLD, size);
+
+  mfem::HypreParVector lhs_par(MPI_COMM_WORLD, size, lhs + start, partitioning.get());
+  mfem::HypreParVector rhs_par(MPI_COMM_WORLD, size, rhs + start, partitioning.get());
+
+  mfem::HypreParVector mfem_result(MPI_COMM_WORLD, size, partitioning.get());
+  add(lhs_par, rhs_par, mfem_result);
+
+  mfem::HypreParVector expr_result(MPI_COMM_WORLD, size, partitioning.get());
+  evaluate(lhs_par + rhs_par, expr_result);
+
+  double* mfem_local = mfem_result;
+  double* expr_local = expr_result;
+
+  for (int i = 0; i < expr_result.Size(); i++) {
+    EXPECT_FLOAT_EQ(mfem_local[i], expr_local[i]);
+  }
+
   MPI_Barrier(MPI_COMM_WORLD);
 }
 

